File: client/src/game/data/colorsInit.ts
import colorsJSON from '../../../../data/colors.json';
import Color from '../types/Color';

// "ColorJSON" is almost exactly the same as "Color"
// However, in "ColorJSON", some fields are optional, but in "Color",
// we want all fields to be initialized
interface ColorJSON {
  name: string;
  fill: string;
  fillColorblind?: string;
  abbreviation?: string;
}

export default function colorsInit() {
  const COLORS = new Map<string, Color>();

  for (const colorJSON of colorsJSON as ColorJSON[]) {
    // Validate the name
    const name: string = colorJSON.name;
    if (name === '') {
      throw new Error('There is a color with an empty name in the "colors.json" file.');
    }

    // Validate the abbreviation
    // If it is not specified, assume that it is the first letter of the color
    const abbreviation: string = colorJSON.abbreviation ?? name.charAt(0);
    if (abbreviation.length !== 1) {
      throw new Error(`The "${colorJSON.name}" color has an abbreviation that is not one letter long.`);
    }

    // Validate the fill
    const fill: string = colorJSON.fill;
    if (colorJSON.fill === '') {
      throw new Error(`The "${colorJSON.name}" color has an empty fill.`);
    }

    // Validate the colorblind fill (which is an alternate fill when "Colorblind Mode" is enabled)
    // If it is not specified, assume that it is the same as the default fill
    const fillColorblind: string = colorJSON.fillColorblind ?? fill;

    // Add it to the map
    const color: Color = {
      name,
      abbreviation,
      fill,
      fillColorblind,
    };
    COLORS.set(colorJSON.name, color);
  }

  return COLORS;
}

File: server/pkg/variants/colors_init.go
package variants

import (
	"encoding/json"
	"io/ioutil"
	"path"
)

func (m *Manager) ColorsInit(dataPath string) {
	// Import the JSON file
	filePath := path.Join(dataPath, "colors.json")
	var fileContents []byte
	if v, err := ioutil.ReadFile(filePath); err != nil {
		m.logger.Fatalf("Failed to read the \"%v\" file: %v", filePath, err)
	} else {
		fileContents = v
	}
	var colorsArray []*Color
	if err := json.Unmarshal(fileContents, &colorsArray); err != nil {
		m.logger.Fatalf("Failed to convert the colors file to JSON: %v", err)
	}

	// Convert the array to a map
	for _, color := range colorsArray {
		// Validate the name
		if color.Name == "" {
			m.logger.Fatal("There is a color with an empty name in the \"colors.json\" file.")
		}

		// Validate that there is an abbreviation
		if color.Abbreviation == "" {
			// Assume that it is the first letter of the color
			color.Abbreviation = string([]rune(color.Name)[0])
		} else if len(color.Abbreviation) != 1 {
			m.logger.Fatalf(
				"The \"%v\" color has an abbreviation that is not one letter long.",
				color.Name,
			)
		}

		// Validate the fill
		if color.Fill == "" {
			m.logger.Fatalf("The \"%v\" color has an empty fill.", color.Name)
		}

		// Validate that all of the names are unique
		if _, ok := m.colors[color.Name]; ok {
			m.logger.Fatalf("There are two colors with the name of: %v", color.Name)
		}

		// Add it to the map
		m.colors[color.Name] = color
	}
}
